pkg/server/server_test.go

```diff
@@ -174,3 +175,142 @@ func (s *TestSuite) TestCRUD(c *C) {
+			})
+		}
+		c.Assert(err, IsNil)
+
+		downloaded := false
+		for j := 0; j < RetryCount; j++ {
+			getResp := bm.Get()
+			if getResp.Status.State == types.DownloadStateDownloaded {
+				downloaded = true
+				break
+			}
+			time.Sleep(RetryInterval)
+		}
+		c.Assert(downloaded, Equals, true)
+
+		err = bm.Delete()
+		c.Assert(err, IsNil)
+	}
+}
+
+func (s *TestSuite) TestSingleBackingImageDownloadCancellation(c *C) {
+	name := "test_downloading_cancellation_backing_image"
+	url := "https://" + name + ".com"
+	uuid := name + "-" + "uuid"
+
+	mockDownloaderFactory := &MockDownloaderFactory{}
+
+	count := 100
+	for i := 0; i < count; i++ {
+		bm := NewBackingImage(name, url, uuid, "/var/lib/longhorn", s.getTestDiskPath(c), mockDownloaderFactory.NewDownloader())
+		bm.SetUpdateChannel(make(chan interface{}, 100))
+		var err error
+
+		err = bm.Delete()
+		c.Assert(err, IsNil)
+
+		if i%2 != 0 {
+			_, err = bm.Pull()
+		} else {
+			_, err = bm.Receive(MockDownloadSize, "SenderAddress", func(portCount int32) (int32, int32, error) {
+				return 0, 0, nil
+			}, func(start, end int32) error {
+				return nil
+			})
+		}
+		c.Assert(err, IsNil)
+
+		isDownloading := false
+		for j := 0; j < 5; j++ {
+			getResp := bm.Get()
+			if getResp.Status.State == types.DownloadStateDownloading && getResp.Status.DownloadProgress > 0 {
+				isDownloading = true
+				break
+			}
+			time.Sleep(RetryInterval)
+		}
+		c.Assert(isDownloading, Equals, true)
+		err = bm.Delete()
+		c.Assert(err, IsNil)
+		getResp := bm.Get()
+		c.Assert(getResp.Status.State, Equals, string(StateFailed))
+	}
+}
+
+func (s *TestSuite) TestBackingImageSimultaneousDownloading(c *C) {
+	name := "test_simultaneous_downloading_backing_image"
+	url := "https://" + name + ".com"
+	uuid := name + "-" + "uuid"
+
+	mockDownloaderFactory := &MockDownloaderFactory{}
+
+	count := 100
+	for i := 0; i < count; i++ {
+		bm := NewBackingImage(name, url, uuid, "/var/lib/longhorn", s.getTestDiskPath(c), mockDownloaderFactory.NewDownloader())
+		bm.SetUpdateChannel(make(chan interface{}, 100))
+
+		err := bm.Delete()
+		c.Assert(err, IsNil)
+
+		// Start to do Pulling and Receiving simultaneously, which is impossible ideally.
+		// Then there must an pull error or sync error.
+		var errPull, errSync error
+		wg := &sync.WaitGroup{}
+		wg.Add(2)
+		go func() {
+			defer wg.Done()
+			_, errPull = bm.Pull()
+		}()
+		go func() {
+			defer wg.Done()
+			_, errSync = bm.Receive(MockDownloadSize, "SenderAddress", func(portCount int32) (int32, int32, error) {
+				return 0, 0, nil
+			}, func(start, end int32) error {
+				return nil
+			})
+		}()
+		wg.Wait()
+		c.Assert((errPull == nil && errSync != nil) || (errPull != nil && errSync == nil), Equals, true)
+
+		isFailed := false
+		for j := 0; j < 5; j++ {
+			getResp := bm.Get()
+			if getResp.Status.State == types.DownloadStateFailed {
+				isFailed = true
+				break
+			}
+			time.Sleep(RetryInterval)
+		}
+		c.Assert(isFailed, Equals, true)
```

Comment:
Don't we expect one or the other operation successful? 

```
c.Assert(isFailed, Equals, false)
```

Reply:
Ok, anwser my question. The following concurrent call will cancel the first one then return w/ an error.

Reply:
For now YES. But I am thinking if we really need to cancel the 1st download when the backing image receives more concurrent download requests.

Reply:
I would suggest no need to do fancy handling, just failed to request if there is an operation running. 

- if one pulling/receiving operation is running, the following requests will be failed until the last one is finished (need to make sure the operation finished finally)